T&A 30750: Poor performance using "Corrections"

lukas-heinrich opened the pull request:
Hi everyone,

This PR addresses the issue described in [30750](https://mantis.ilias.de/view.php?id=30750) where the corrections process for adding an answer option in ILIAS is very slow. With this PR, I propose several changes to improve the process performance.

For reproduction, I worked with a test containing 11 questions (10 of them being Cloze Questions) that was completed by 500 participants to achieve approximately the same scale as described in the ticket and to conduct a better comparison.

As can be traced in the commit history, I implemented and tested each change separately step by step. Using comparison checks, I verified that my changes do not alter the expected processing results. This PR includes the following changes:

1. The "Add Answer" operation always explicitly refers to a specific question. However, during point recalculation, every question in the test is updated in the database. I could not identify any scenario where recalculating all questions affects the overall result calculation.
2. Updating the total results (*tst_result_cache*) only after the final recalculation instead of after each update
3. Creation of question objects once; I found that this method is called for each test attempt and rebuilds the same data each time. As far as I can tell, the object has no user state, so it's possible to create the object once and reuse it subsequently
4. Optimized method for updating points and querying all participant results only calculates information relevant to the post-correction process, thereby drastically reducing database queries
5. Optimization of database queries by caching parameters for logging and caching combinations for Cloze-Gaps


Through analysis using external tools, I measured the performance as follows:

- Current status in `release_9`: 103 seconds, ~235,000 database queries, 570MB peak memory
- Status after the last commit in this PR: 1.8 seconds, ~12,000 database queries, 52MB peak memory

The changes and measurement results have been documented in more detail in internal documentation. We can provide this document if needed.

Thank you in advance for the review, I look forward to feedback on these changes.

Best,
@lukas-heinrich

lukas-heinrich committed:
fix: provide question id to perform the calculations for the corrected question only

lukas-heinrich committed:
fix: remove duplicate result cache update as it was already written in \assQuestion::_updateTestPassResults

lukas-heinrich committed:
fix: temporary store question object instead of creating a new one for every pass

lukas-heinrich committed:
refactor: replace assQuestion::_setReachedPoints by optimized method which only updates database if necessary

lukas-heinrich committed:
refactor: use already existing parameter to avoid checking if cloze combination exists on every iteration

lukas-heinrich committed:
refactor: use ilObjAssessmentFolder::_addLog directly to avoid test id lookups

lukas-heinrich committed:
refactor: pass through already loaded passes to avoid multiple loading

lukas-heinrich committed:
refactor: split \ilTestEvaluationFactory::getEvaluationData to only query needed result data

lukas-heinrich committed:
fix: provide global ilUser in test cases

dsstrassner commented:
Hi @lukas-heinrich, please provide the document to @mjansenDatabay & @kergomard & me via e-mail.

mjansenDatabay commented on Modules/Test/classes/class.ilTestCorrectionsGUI.php:
```diff
@@ -286,10 +286,12 @@ protected function showSolution()
     }
 
-    protected function showAnswerStatistic()
+    protected function showAnswerStatistic($participant_results = null)
```
```suggestion
    /**
     * @param null|array<int, ilTestEvaluationUserData> $participant_results
     */
    protected function showAnswerStatistic(?array $participant_results = null)
```

lukas-heinrich committed:
refactor: add missing parameter type and docblock

lukas-heinrich committed:
fix: use correct null coalescing operator

This commit also fixes a problem where the cache was not updated correctly. This results in an incorrect score being displayed in the results table. The points from the database and in the detailed view are not affected by this problem.

kergomard reviewed (changes requested):
Thank you very much for the PR a for tackling the performance issue @lukas-heinrich !

I've the following immediate concerns:
- [x] As far as I can see `ilTestScoring::recalculateSolution()` is still needed in `ilTestScoringGUI::saveManScoringParticipantScreen()` and cannot be removed.

Some small things:
-  [x] Please do not remove typing like [here](https://github.com/ILIAS-eLearning/ILIAS/pull/8978/files#diff-4e9b80468abe4288bf99750d58dd525cc4a2824f679b9d957e9e4c33fa163072R136). If there is a problem it is clearly not the typing, but the parameters passed.
- [x] Please restructure this if - else. It looks bonkers.
- [x] Please use consistent doc-blocks for arrays of the type `array<type>` (or `list<type>`).
- [x] [This check](https://github.com/ILIAS-eLearning/ILIAS/pull/8978/files#diff-4e9b80468abe4288bf99750d58dd525cc4a2824f679b9d957e9e4c33fa163072R126) is useless: We always return an array. Please remove it.
- [x] Please remove [the intermediate variable here](https://github.com/ILIAS-eLearning/ILIAS/pull/8978/files#diff-4e9b80468abe4288bf99750d58dd525cc4a2824f679b9d957e9e4c33fa163072R128) and use strict typing below. We do know what type this variable has, don't we? If we don't: Please fix the problem where it originates.
- [x] Please remove [the intermediate variable here](https://github.com/ILIAS-eLearning/ILIAS/pull/8978/files#diff-4e9b80468abe4288bf99750d58dd525cc4a2824f679b9d957e9e4c33fa163072R144). You can keep things readable by using line breaks in the following line.
- [x] Please add a return-type to `ilTestScoring::updateReachedPoints` and use strict comparisons within it.
- [x] Single Quotes are enough [here](https://github.com/ILIAS-eLearning/ILIAS/pull/8978/files#diff-4e9b80468abe4288bf99750d58dd525cc4a2824f679b9d957e9e4c33fa163072R197).
- [x] As far as I can see the float-cast [here](https://github.com/ILIAS-eLearning/ILIAS/pull/8978/files#diff-4e9b80468abe4288bf99750d58dd525cc4a2824f679b9d957e9e4c33fa163072R201) can go.

Please create a PR for this against ILIAS 10, with the following changes. I might have missed something, so all of this is conditioned on it being possible.
- [ ] Remove `assQuestion::_setReachedpoints()` completely.
- [ ] Remove the duplication of code in `ilTestEvaluationFactory::getEvaluationData()` and `ilTestEvaluationFactory::getCorrectionsEvaluationData()` by restructuring the functions correspondingly.
- [ ] Make the parameter in `ilTestCorrectionsGUI::showAnswerStatistic()` compulsory (maybe you also need to add one more that hands in the question-object). If it isn't possible, it isn't, but I deeply dislike this type of optional parameters. From where I stand they are a smell.
- [ ] Make sure to not introduce any dependence on the `$DIC` in `TestScoring`.

Thanks again for all the work! I did not test the functioning of it, but I will give it a quick spin once we have the PR for ILIAS 10.

Thanks again and best,
@kergomard

lukas-heinrich committed:
refactor: improve code quality and readability by using strict types and restructure if-statements

lukas-heinrich commented:
Hi @kergomard  ,

thank you very much for the review and the detailed feedback. I am preparing a PR for Ilias 10, I have just pushed the proposed changes for the release_9.

Best
@lukas-heinrich

kergomard commented:
Thank you very much again @lukas-heinrich !

I will wait for the PR for ILIAS 10 and merge all in one, ...just for me not loosing the overview.

There is one more little thing I missed before, sorry: Could you have a look at the failing Content-Style-Checks?

Thanks and best,
@kergomard

lukas-heinrich committed:
refactor: cs fix

rfalkenstein commented:
Hi guys, we are currently testing this patch for ILIAS 8 and it works well. The loading times for adding a new answer in the corrections tab was reduced from 8 minutes to 3 seconds. In the next step we could offer to introduce these fixes into our ILIAS 9 staging installation. 
@lukas-heinrich @thojou  do you consider the current state of the branch "bugfix/9/30750" ready for testing with ILIAS 9? Then we'll introduce this into our ILIAS 9 staging installation and test (again, but for ILIAS 9) against tests with a lot of users.

thojou commented:
Hey @rfalkenstein,

Thanks for your feedback and for testing the changes in ILIAS 8.
This branch is complete as well. 
We’d really appreciate if you could test these changes on your ILIAS 9 staging installation.

Best,
@thojou

lukas-heinrich committed:
fix: Argument #4 ($title) must be of type string by creating a optimized buildEvaluationData method

lukas-heinrich commented:
Hi @kergomard ,

I had to push another commit on this branch for Release 9 and rebase it. The background is that the `getEvaluationData` method has been changed in the meantime. To prevent changes to the method from being implemented in two places in the future, I have restructured and refactored the method call.

I would appreciate if you could take another look at the latest commit. We will provide a PR for Ilias 10 soon.

Thank you very much and best
Lukas

kergomard committed:
Test: Restructure Evaluation Factory

kergomard commented:
Hi @lukas-heinrich 

Thank you very much for your changes.

I'm going to be honest: I don't like this pattern at all:
- It doesn't simplify, but makes the code more convoluted.
- The function is still far too long.
- We have to "adaptor"-functions that just add an indirection.

I decided to quickly give it a go, as I think it is easier to demonstrate in what direction I would go, then to try to explain it. I couldn't create a PR against your branch (don't ask me why), but you can find a proposal here: https://github.com/kergomard/ILIAS/commit/9bd9068c40d3774aac1eaccab4e2052faebfb9fc . If you decide to use this: Please check it thoroughly! I think it works and I clicked on all buttons on the surface, but I didn't test it thoroughly.

Thanks again and best,
@kergomard

lukas-heinrich commented:
Hi @kergomard ,

thank you for the feedback and the implementation. I agree, this one is a much more elegant solution for structuring the methods. I have pushed the changes from your commit and thoroughly tested them - I haven't found any issues.

Thanks again and Best
@lukas-heinrich

lukas-heinrich committed:
fix: add all answered questions to attempt

kergomard commented:
Thank you for catching this error @lukas-heinrich !

Just that we are not blocking one another here: It is still correct that I'm waiting for a PR against ILIAS 10 here, right?

And one very small  request (at least for ILIAS 10):
- [ ] Could you please remove the reassignment to `$attempt` and pass `$this->addQuestionToAttempt($attempt, $row)` directly to `$user_eval_data->addPass()` (`$user_eval_data->addPass($row['pass'], $this->addQuestionToAttempt($attempt, $row))` - you need to split it on multiple lines, though). I do really believe we should treat our variables in function-scope as constants as much as we can (maybe one day we will be able to declare them as such).

Best,
@kergomard

lukas-heinrich commented:
Hi @kergomard ,

thank you for the hint, I will adjust it accordingly. 
It is correct that the revision for Ilias 10 is still in progress. However, I assume that this will be completed by the end of this week and that the PR can be provided after thorough testing.

Best
@lukas-heinrich

kergomard commented:
Hi @lukas-heinrich 

The PR against ILIAS 10 is still missing, right?

Thank you very much and best,
@kergomard

thojou commented:
Hey @kergomard,

We are currently conducting internal reviews on the ILIAS 10 PR.
As soon as we're ready, we will publish the branch.

Best,
@thojou

lukas-heinrich commented:
Hi @kergomard 

The PR against ILIAS 10 is now ready: #9277 

Best,
@lukas-heinrich

thojou commented:
Hey,

We've also received feedback from testing this branch on production datasets. No further issues were reported. From our perspective, this PR is ready to go! 🙂

Best,
@thojou

kergomard commented:
Thank you very much @lukas-heinrich 

As we now have the corresponding PR for ILIAS 10, I've merged this.

Best,
@kergomard
